projects/ngx-quill/src/lib/quill-editor.component.ts

```diff
@@ -615,6 +622,22 @@ export abstract class QuillEditorBase implements AfterViewInit, ControlValueAcce
 
     return valid ? null : err
   }
+
+  private debounce<T extends(...args: any[]) => any>(callback: T): ((...args: Parameters<T>) => void) {
+    let timer: number
```

Comment:
i think this should be a class property, because we need to clear the timeout when the component gets destroyed, as well

Reply:
Done, I created an array to keep the different IDs (?) to destroy.

Reply:
eehhm. in general you do not need to cache a list of timers i think. and you do not need the `let timer: number` in the function.

Just create a class property `changeDebounceTimer: number | null = null`.
And use it here:

```
if (this.changeDebounceTimer !== null) {
  this.doc.defaultView.clearTimeout(this.changeDebounceTimer)
  this.changeDebounceTimer = null
}
``` 
So we clear the last timeout when the debounce is called again. so there is only one running/active timeout
And in ngOnDestroy you do the same thing:

```
if (this.changeDebounceTimer !== null) {
  this.doc.defaultView.clearTimeout(changeDebounceTimer)
  this.changeDebounceTimer = null
}
```

Reply:
the whole thing would be something like:

```TS
  changeDebounceTimer: number | null = null
  ...

  ngOnDestroy() {
    if (this.quillEditor) {
      ...
      if (this.debounceTime !== null) {
        this.doc.defaultView.clearTimeout(this.changeDebounceTimer)
      }
    }
  }

  private debounce<T extends(...args: any[]) => any>(callback: T): ((...args: Parameters<T>) => void) {
    return (...args: Parameters<T>) => {
      if (this.debounceTime !== null) {
        callback(...args)
        return
      }

      this.doc.defaultView.clearTimeout(this.changeDebounceTimer)
      this.changeDebounceTimer = null

      this.changeDebounceTimer = this.doc.defaultView.setTimeout(() => {
        callback(...args)
      }, this.debounceTime)
    }
  }
```

Reply:
Couldn't it cause issues that multiple debounced methods rely on one single timer? (both `editorChangeHandler` and `textChangeHandler`)
I apologize if I'm missing something, but your implentation seems to fail the tests because the second debounced method always "overrides"/clears the timer of the previous, yet different method.

Reply:
ahh i just did not checked that we use it twice. but maybe we can only store 2 timeouts or as many as needed.
e.g.
if a timeout is cleared in the debounce function we should remove it from the array

Reply:
```
this.doc.defaultView.clearTimeout(timer)
this.debounceTimers = this.debounceTimers.filter((debounceTimer) => debounceTimer !== timer)
```
That was my original intention with the `filter`. As far as I've seen this removes the cleared timer from the array which means that the most elements the array will have is 2.
In the `ngOnDestroy` I didn't clear the array, but I guess that's rather pointless because the component gets destroyed anyways.

Reply:
but destroying the component does not cancel the timeout, because timeouts are not connected to dom nodes so it is useful to clean them up ;-)

Reply:
Didn't I clean them up by doing this? ([commit](https://github.com/KillerCodeMonkey/ngx-quill/pull/1226/commits/f97e05578e51a5a43a019580f010f08e43498c34#diff-e7777fef0077454e16b00b915eb03114f53673cbd5d5e966d1f45c3bbbbca151R456) - line 456)
```
ngOnDestroy() {
  ...
  this.debounceTimers.forEach((timer) => this.doc.defaultView.clearTimeout(timer))
}
```

Anyways, we've noticed that the finished timers are not removed from the array which could result in an infinite large array (eg. if the debounced function is called only X times then the last timer will remain in the array even though it's finished), therefore I fixed that and added some more tests.
I also added a test that makes sure that the timers are cancelled once the component gets destroyed.

I'm sorry if I'm still missing something, but I guess now everything is supposed to be cancelled AND the `debounceTimers` array is correctly kept in sync with the active timers.